include/crocoddyl/core/activations/quadratic-flat-exp.hpp

```diff
@@ -0,0 +1,156 @@
+///////////////////////////////////////////////////////////////////////////////
+// BSD 3-Clause License
+//
+// Copyright (C) 2020, LAAS-CNRS, University of Edinburgh
+// Copyright note valid unless otherwise stated in individual files.
+// All rights reserved.
+///////////////////////////////////////////////////////////////////////////////
+
+#ifndef CROCODDYL_CORE_ACTIVATIONS_QUADRATIC_FLAT_EXP_HPP_
+#define CROCODDYL_CORE_ACTIVATIONS_QUADRATIC_FLAT_EXP_HPP_
+
+#include "crocoddyl/core/activation-base.hpp"
+#include "crocoddyl/core/fwd.hpp"
+#include "crocoddyl/core/utils/exception.hpp"
+#include <iostream>
+#include <stdexcept>
+
+namespace crocoddyl {
+
+/*
+ * @brief Quadratic-flat-exp activation
+ *
+ * This activation function describes a quadratic exponential activation
+ * depending on the square norm of a residual vector, i.e. \f[ \begin{equation}
+ * 1 - exp(-\|\mathbf{r}\|^2 / \alpha) \end{equation} \f] where \f$\alpha\f$
+ * defines the width of the quadratic basin, \f$r\f$ is the scalar residual,
+ * \f$nr\f$ is the dimension of the residual vector. Far
+ * away from zero, the quadFlat activation is nearly flat.
+ *
+ * The computation of the function and it derivatives are carried out in
+ * `calc()` and `caldDiff()`, respectively.
+ *
+ * \sa `calc()`, `calcDiff()`, `createData()`
+ */
+template <typename _Scalar>
+class ActivationModelQuadFlatExpTpl
+    : public ActivationModelAbstractTpl<_Scalar> {
+public:
+  EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+
+  typedef _Scalar Scalar;
+  typedef MathBaseTpl<Scalar> MathBase;
+  typedef ActivationModelAbstractTpl<Scalar> Base;
+  typedef ActivationDataAbstractTpl<Scalar> ActivationDataAbstract;
+  typedef ActivationDataQuadFlatExpTpl<Scalar> Data;
+  typedef typename MathBase::VectorXs VectorXs;
+  typedef typename MathBase::MatrixXs MatrixXs;
+
+  /*
+   * @brief Initialize the quadFlatExp activation model
+   *
+   * The default `alpha` value is defined as 1.
+   *
+   * @param[in] nr     Dimension of the residual vector
+   * @param[in] alpha  Width of quadratic basin (default: 1.)
+   */
+
+  explicit ActivationModelQuadFlatExpTpl(const std::size_t &nr,
```

Comment:
Let's do not use `const reference` for primitives (the primitives are `std::size_t`, `double`, `Scalar` or `bool`.
Thus the code should be modified to
```cpp
   explicit ActivationModelQuadFlatExpTpl(std::size_t nr, Scalar alpha = Scalar(1.))
```

Reply:
You can use const here. It will be helpful to check whether the quantity should be modified or not.

Reply:
The quantity should be never modified, that should be clear for any developer.
Do you really think we need to care about it?

Reply:
If you know that the quantities should not be modified, then you should mark it explicitly.
It may also help the compiler to optimize the code accordingly depending on the context.